SleepNumberController_Driver.groovy

```diff
@@ -139,6 +143,88 @@ def updated() {
   poll()
 }
 
+def uninstalled() {
+   deleteChildDevice(getChildDNI("outlet"))
+   deleteChildDevice(getChildDNI("underbedlight"))
+}
+
+def manageChildren() {
+    switch (state.type) {
+        case "presence":
+            def outlet = getOutlet()
+            if (!outlet) {
+                outlet = addChildDevice('hubitat','Generic Component Switch', getChildDNI("outlet"),
+                                        [ label: getChildName("Outlet"),
+                                         componentName: "outlet",
+                                         componentLabel: getChildName("Outlet"),                	                  
+                                         completedSetup:true])
+                log.info("Created Outlet child device")
+            }
+            def underbed = getUnderbed()
+            if (!underbed) {
+                outlet = addChildDevice('hubitat','Generic Component Switch', getChildDNI("underbedlight"),
+                                        [ label: getChildName("Underbed Light"),
+                                         componentName: "underbedlight",
+                                         componentLabel: getChildName("Underbed Light"),                	                  
+                                         completedSetup:true])
+                log.info("Created Underbed light child device")
+            }
+            break
+    }
+}
+
+def getOutlet() {
+    return childDevices.find({it.deviceNetworkId == getChildDNI("outlet")})
+}
+
+def getUnderbed() {
+    return childDevices.find({it.deviceNetworkId == getChildDNI("underbedlight")})
+}
+
+
+def getChildName(name) {
+    return "Sleep Number ${state.side} - ${name}"
+}
+
+def getChildDNI(name) {
+	return device.deviceNetworkId + "-" + name
+}
+
+def getChildDNI(type, name) {
+    return device.deviceNetworkId + + "-${type}-${name}"
+}
+
+def componentRefresh(device) {
+    log.warn("Component Refresh not implemented")
+}
+
+def componentOn(device) {    
```

Comment:
(another naive question) - is this called when the child device is turned on?

Reply:
Yes the generic component devices define a standard API for calls back to the parent.  This is not well documented as far as I know, but you can see the code here:

https://github.com/hubitat/HubitatPublic/blob/master/examples/drivers/componentSwitch.groovy